File: server/models/user.model.ts
import jwt from "jsonwebtoken";
import mongoose from "mongoose";
import { v4 as uuidv4 } from "uuid";
import { hashPassword, verifyPassword } from "../utils/password";

interface matchType {
  [key: string]: number;
}

export interface UserInput extends mongoose.Document {
  user_id: string;
  first_name: string;
  email: string;
  password: string;
  dob_day: number;
  dob_month: number;
  dob_year: number;
  show_gender: boolean;
  gender_identity: string;
  gender_interest: string;
  interest: string[];
  url1: string;
  url2: string;
  url3: string;
  about: string;
  isVerified: Boolean;
  isUpdated: Boolean;
  matches: matchType;
  chatList: string[];
  pending: string[];
  coords: number[];
  key: string;
  refreshToken: string;
  pfp: { data: string; contentType: string };

  createJWT(): string;
  createRJWT(): string;
  checkPassword(password: string): boolean;
}

const UserSchema = new mongoose.Schema({
  user_id: {
    type: String,
    default: uuidv4(),
  },
  first_name: {
    type: String,
    minlenght: 3,
  },
  email: {
    type: String,
    required: [true, "Email is required!"],
    match: [
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
      "Provide a valid email!",
    ],
    unique: true,
  },
  password: {
    type: String,
    required: [true, "Password is required!"],
    minlength: 6,
  },
  dob_day: {
    type: Number,
  },
  dob_month: { type: Number },
  dob_year: { type: Number },
  show_gender: { type: Boolean, default: false },
  gender_identity: {
    type: String,
    enum: {
      values: ["male", "female", "others"],
      message: "{VALUE} is not supported! ",
    },
  },
  gender_interest: { type: String },
  interest: {
    type: Array,
    default: [],
  },
  pfp: {
    data: String,
    contentType: String,
  },
  url1: { type: String },
  url2: { type: String },
  url3: { type: String },
  about: { type: String },
  matches: { type: Object, default: {} },
  isVerified: {
    type: Boolean,
    default: false,
  },
  isUpdated: {
    type: Boolean,
    default: false,
  },
  chatList: {
    type: Array,
    default: [],
  },
  pending: {
    type: Array,
    default: [],
  },
  coord: {
    type: Array,
    default: [27.7172, 85.324],
  },
  key: { type: String },
  refreshToken: String,
});

//pre middleware - hashing
UserSchema.pre("save", async function () {
  this.password = await hashPassword(this.password);
});

//instance methods

//access token
UserSchema.methods.createJWT = function (this: UserInput) {
  return jwt.sign(
    { userID: this._id, email: this.email },
    process.env.JWT_SECRET as string,
    {
      expiresIn: process.env.JWT_LIFETIME,
    }
  );
};

//refresh token
UserSchema.methods.createRJWT = function (this: UserInput) {
  return jwt.sign(
    { userID: this._id, email: this.email },
    process.env.RJWT_SECRET as string,
    {
      expiresIn: process.env.RJWT_LIFETIME,
    }
  );
};

UserSchema.methods.checkPassword = async function (
  password: string
): Promise<boolean> {
  return await verifyPassword(password, this.password);
};

export default mongoose.model<UserInput>("User", UserSchema);

File: client/lib/services/models/user_model.dart
// To parse this JSON data, do
//
//     final userModel = userModelFromJson(jsonString);

import 'dart:convert';

List<UserModel> userModelFromJson(String str) =>
    List<UserModel>.from(json.decode(str).map((x) => UserModel.fromJson(x)));

String userModelToJson(List<UserModel> data) =>
    json.encode(List<dynamic>.from(data.map((x) => x.toJson())));

class UserModel {
  String? id;
  String? userId;
  String? hashedPassword;
  String? firstName;
  int? dobDay;
  int? dobMonth;
  int? dobYear;
  bool? showGender;
  String? genderIdentity;
  String? genderInterest;
  String? email;
  String? url1;
  String? url2;
  String? about;
  List<Match>? matches;
  String? password;
  int? v;
  String? url;

  UserModel({
    this.id,
    this.userId,
    this.hashedPassword,
    this.firstName,
    this.dobDay,
    this.dobMonth,
    this.dobYear,
    this.showGender,
    this.genderIdentity,
    this.genderInterest,
    this.email,
    this.url1,
    this.url2,
    this.about,
    this.matches,
    this.password,
    this.v,
    this.url,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) => UserModel(
        id: json["_id"],
        userId: json["user_id"],
        hashedPassword: json["hashed_password"],
        firstName: json["first_name"],
        dobDay: json["dob_day"],
        dobMonth: json["dob_month"],
        dobYear: json["dob_year"],
        showGender: json["show_gender"],
        genderIdentity: json["gender_identity"],
        genderInterest: json["gender_interest"],
        email: json["email"],
        url1: json["url1"],
        url2: json["url2"],
        about: json["about"],
        matches: json["matches"] == null
            ? []
            : List<Match>.from(json["matches"]!.map((x) => Match.fromJson(x))),
        password: json["password"],
        v: json["__v"],
        url: json["url"],
      );

  Map<String, dynamic> toJson() => {
        "_id": id,
        "user_id": userId,
        "hashed_password": hashedPassword,
        "first_name": firstName,
        "dob_day": dobDay,
        "dob_month": dobMonth,
        "dob_year": dobYear,
        "show_gender": showGender,
        "gender_identity": genderIdentity,
        "gender_interest": genderInterest,
        "email": email,
        "url1": url1,
        "url2": url2,
        "about": about,
        "matches": matches == null
            ? []
            : List<dynamic>.from(matches!.map((x) => x.toJson())),
        "password": password,
        "__v": v,
        "url": url,
      };
}

class Match {
  String? userId;

  Match({
    this.userId,
  });

  factory Match.fromJson(Map<String, dynamic> json) => Match(
        userId: json["user_id"],
      );

  Map<String, dynamic> toJson() => {
        "user_id": userId,
      };
}
